GithubRunnerRegistration/GithubRunnerRegistration/Register.cs

```diff
@@ -0,0 +1,116 @@
+
+        public string RunnerRegistrationToken { get; set; }
+
+        public Register(RegistrationRequest request, string binary = "/actions-runner/bin/Runner.Listener.dll")
+        {
+            Array.ForEach(request.Labels, this.validateString);
+            this.request = request;
+            this.binary = new FileInfo(binary);
+        }
+
+        private GitHubClient getClient() => new GitHubClient(new ProductHeaderValue(nameof(GithubRunnerRegistration))) { Credentials = new Credentials(this.request.AdminPat) };
+
+        private void validateString(string str)
+        {
+            var rx = "[^0-9a-z-_]";
+            if (Regex.IsMatch(str, rx, RegexOptions.IgnoreCase))
+            {
+                throw new InvalidOperationException($"'{str}' contains invalid chars (rx: {rx})");
+            }
+        }
+
+        public async Task Setup(IGitHubClient clnt = default)
+        {
+            var client = clnt != default ? clnt : this.getClient();
+            Repository repository;
+            try
+            {
+                repository = await client.Repository.Get(this.request.Owner, this.request.Repository);
+            }
+            catch (Exception ex)
+            {
+                throw new SetupException("Error getting repository", ex);
+            }
+            if (!repository.Permissions.Admin)
+            {
+                throw new SetupException("Not admin", default);
+            }
+            var tokenResult = await client.Connection.Post<TokenResult>(new Uri($"https://api.github.com/repos/{this.request.Owner}/{this.request.Repository}/actions/runners/registration-token"));
+            this.RunnerRegistrationToken = tokenResult.Body.Token;
+        }
+
+        public async Task<Dictionary<string,string>> AddRunner(string name) => await this.RegisterRunner(name, new GetCredentials(), this.Run);
+
+        private void Run(ProcessStartInfo arg)
+        {
+            var proc = Process.Start(arg);
+            proc.WaitForExit();
+            if (proc.ExitCode != 0)
+            {
+                throw new Exception("Binary returned code " + proc.ExitCode);
+            }
+        }
+
+        public async Task<Dictionary<string, string>> RegisterRunner(string name, IGetCredentials getCreds, Action<ProcessStartInfo> run)
+        {
+            this.validateString(name);
+            var tmp = Path.Combine(Path.GetTempPath(), name);
+            if (Directory.Exists(tmp))
+            {
+                Directory.Delete(tmp, true);
+            }
+            Directory.CreateDirectory(tmp);
+            DirectoryCopy(binary.Directory.FullName, Path.Combine(tmp, "runner"));
+
+            var startInfo = new ProcessStartInfo("dotnet", $"{this.binary.Name} configure --name \"{name}\" --token {this.RunnerRegistrationToken} --url \"https://github.com/{this.request.Owner}/{this.request.Repository}\" --labels \"{string.Join(",", this.request.Labels)}\" --replace --unattended");

```

Comment:
Why are we calling a dotnet process as a subprocess rather than binding in and calling the methods directly? Seems super weird at a first reading

Reply:
Because the library gets updated frequently